src/servers/buttongestureadapter.ts

```diff
@@ -0,0 +1,95 @@
+import {
+    ButtonEvent,
+    ButtonGestureEvent,
+    EVENT,
+    SRV_BUTTON_GESTURE,
+} from "../jdom/constants"
+import SensorServer from "./sensorserver"
+import { JDEvent, JDService } from "../jdom/jacdac-jdom"
+
+export default class ButtonGestureAdapter extends SensorServer<[number]> {
+    protected state: "up" | "down_click" | "up_click" | "down_held" = "up"
+
+    // TODO this is only used to gate timeouts using threading + wait.
+    // A better solution would be to terminate threads so they don't fire when they're obsoleted.
+    // TODO if keeping this architecture, perhaps this needs to be more specific, eg clickTimeoutCounter
+    protected eventCounter = 0
+
+    // counter to detect double-clicks and the like
+    // reset upon a event being generated
+    protected clickCounter = 0
+
+    static readonly clickTimeoutMs: number = 200
+    static readonly multiClickTimeoutMs: number = 200
+
+    constructor(button: JDService, instanceName?: string) {
+        // TODO should this take not a service so it can be instantiated before a button is announced on the bus?
+        // (to avoid the async boilerplate nightmare)
+        super(SRV_BUTTON_GESTURE, {
+            instanceName,
+        })
+
+        button.on(EVENT, (evs: JDEvent[]) => {
+            evs.forEach((ev) => {
+                const now = this.device.bus.timestamp
+                if (ev.code == ButtonEvent.Down) {
+                    this.onSourceButtonDown()
+                } else if (ev.code == ButtonEvent.Up) {
+                    this.onSourceButtonUp()
```

Comment:
I don't know if this is yet exposed, but up events contain a timestamp of the time they were down.

Reply:
Might make click detection easier...